gpclient/gpserviceaccount.py

```diff
@@ -52,56 +73,95 @@ class GPServiceAccount():
+                    credentialsSet = True
+                    logging.info("""using user defined environment variables to
+                        create GPServiceAccount""")
+                elif apiKey:
+                    self.__setIamCreds__(url, instanceId, apiKey)
+                    credentialsSet = True
+                    logging.info("""using user defined environment variables to
+                                    create GPServiceAccount supporting IAM auth.""")
         if not credentialsSet and serviceInstanceName is not None:
-                (url, instanceId, userId, password) = \
+            (url, instanceId, userId, password, apiKey) = \
                     self.__parse_vcap_services_env_var(serviceInstanceName)
-                if url and instanceId and userId and password:
+            if url and instanceId:
+                if userId and password:
                     self.__setcreds__(url, instanceId, userId, password)
                     credentialsSet = True
                     logging.info("""using VCAP_SERVICES environment variable to
-                        create GPServiceAccount""")
+                            create GPServiceAccount""")
+                elif apiKey:
+                    self.__setIamCreds__(url, instanceId, apiKey)
+                    credentialsSet = True
+                    logging.info("""using VCAP_SERVICES environment variable to
+                            create GPServiceAccount supporting IAM auth.""")
         if not credentialsSet and credentialsJson is not None:
-            (url, instanceId, userId, password) = self.__get_credentials_from_file(credentialsJson)
-            if url and instanceId and userId and password:
-                self.__setcreds__(url, instanceId, userId, password)
-                credentialsSet = True
-                logging.info("""using user defined environment variables to
-                    create GPServiceAccount""")
+            (url, instanceId, userId, password, apiKey) = self.__get_credentials_from_file(credentialsJson)
+            if url and instanceId:
+                if userId and password:
+                    self.__setcreds__(url, instanceId, userId, password)
+                    credentialsSet = True
+                    logging.info("""using user defined environment variables to
+                        create GPServiceAccount""")
+                elif apiKey:
+                    self.__setcreds__(url, instanceId, apiKey)
+                    credentialsSet = True
+                    logging.info("""using user defined environment variables to
+                        create GPServiceAccount supporting IAM auth.""")
         # make sure that all the vars are set
-        assert self.__url, ('url is not a string: <%s>')
+        assert self.__url, ('url is not a string: <%s>' % self.__url)
         assert self.__instanceId, ('instanceId is not a string: <%s>')

```

Comment:
same here, needs `% self.__instanceId` ?

Reply:
No string will ever evaluate to false in here. I think original author's intent was to inform if this assert fails, say this should be string type, not really familiar with the conventions in py. Does seem misleading - agree.

Only following values of variable will lead assert to fail:
None
False
zero of any numeric type, for example, 0, 0L, 0.0, 0j.
any empty sequence, for example, '', (), [].
any empty mapping, for example, {}.
instances of user-defined classes, if the class defines a __nonzero__() or __len__() method, when that method returns the integer zero or bool value False.
src: https://docs.python.org/2.4/lib/truth.html

Reply:
hmm ok, tried something in py console, should do %... for all to print out the invalid values